File: frontend/src/types/navigation.ts
export interface BoundingBox {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface CandidateElement {
  label: string;
  bbox: BoundingBox;
  confidence: number;
  is_chosen: boolean;
}

export interface DetectedElement {
  label: string;
  element_type: string;
  bbox: BoundingBox;
  confidence: number;
  interactable: boolean;
  accessibility_issues: string[];
}

export interface PerceptionResult {
  page_description: string;
  page_purpose: string;
  has_modal_or_overlay: boolean;
  loading_state: boolean;
  elements: DetectedElement[];
}

export type ActionType = "click" | "type" | "scroll_down" | "scroll_up" | "press_key" | "hover" | "wait" | "go_back" | "done";

export type FrictionSeverity = "low" | "medium" | "high" | "critical";

export interface ActionPlan {
  action_type: ActionType;
  target_element: string | null;
  coordinates: [number, number] | null;
  input_text: string | null;
  key: string | null;
  seconds: number | null;
  reasoning: string;
  confidence: number;
  candidates: CandidateElement[];
}

export type FrictionCategory = "navigation" | "contrast" | "affordance" | "copy" | "error" | "performance" | "accessibility";

export interface CategorizedFriction {
  category: FrictionCategory;
  description: string;
  severity: "low" | "medium" | "high" | "critical";
  evidence: string;
}

export interface EvaluationResult {
  progress_made: boolean;
  goal_achieved: boolean;
  page_changed: boolean;
  description: string;
  friction_detected: string[];
  friction_items?: CategorizedFriction[];
  confidence: number;
}

export interface ActionResult {
  success: boolean;
  url_after: string;
  error: string | null;
}

// Matches backend NavigationStep schema exactly
export interface NavigationStep {
  step_number: number;
  perception: PerceptionResult;
  plan: ActionPlan;
  action_result: ActionResult;
  evaluation: EvaluationResult;
  screenshot_before_b64: string;
  screenshot_after_b64: string;
  screenshot_before_url?: string;
  screenshot_after_url?: string;
  timestamp: string;
  is_recovery: boolean;
  recovery_reason: string | null;
}

// Real-time thought events from the PPAE loop
export interface ThoughtEvent {
  type: "thought";
  phase: "perceive" | "plan" | "act" | "evaluate";
  message: string;
  data?: Record<string, unknown>;
}

export interface NavigationMetrics {
  total_steps: number;
  retries: number;
  backtracks: number;
  time_seconds: number;
  friction_count: number;
  outcome: string;
  recovery_steps: number;
}

export interface SessionData {
  id: string;
  url: string;
  goal: string;
  persona: string | null;
  status: string;
  total_steps: number;
  completed: boolean;
}

// Matches backend PersonaConfig schema
export interface PersonaConfig {
  key: string;
  name: string;
  description: string;
  prompt_suffix: string;
  evaluation_suffix: string;
  behavior_modifiers: Record<string, unknown>;
}

export interface PathStep {
  step: number;
  action: string;
  target: string | null;
  friction: boolean;
  url: string;
}

export interface PersonaRunResult {
  persona_key: string;
  persona_name: string;
  total_steps: number;
  retries: number;
  backtracks: number;
  time_seconds: number;
  friction_count: number;
  friction_by_category: Record<string, number>;
  friction_score: number;
  ux_risk_index: number;
  outcome: string;
  session_id: string;
  path: PathStep[];
  verdict: string;
}

export interface ComparisonResult {
  url: string;
  goal: string;
  results: PersonaRunResult[];
}

// Matches backend FrictionItem schema
export interface FrictionItem {
  category: FrictionCategory;
  description: string;
  severity: FrictionSeverity;
  evidence_step: number;
  evidence_screenshot_url: string;
  improvement_suggestion: string;
  persona_impacted: string | null;
}

// Matches backend StepSummary schema
export interface StepSummary {
  step_number: number;
  action_type: string;
  target: string | null;
  reasoning: string;
  friction_detected: string[];
  confidence: number;
  screenshot_url: string;
}

export interface ErrorClassification {
  error_type: "none" | "network" | "timeout" | "blocked" | "login_wall" | "safety" | "unknown";
  details: string;
  recoverable: boolean;
}

// Matches backend FrictionReport schema
export interface FrictionReport {
  session_id: string;
  url: string;
  goal: string;
  persona: string | null;
  status: string;
  total_steps: number;
  total_time_seconds: number;
  friction_items: FrictionItem[];
  friction_score: number;
  ux_risk_index: number;
  metrics: Record<string, number>;
  error_classification: ErrorClassification;
  executive_summary: string;
  improvement_priorities: string[];
  step_timeline: StepSummary[];
}

File: backend/schemas/navigation.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class BoundingBox(BaseModel):
    x: int = Field(description="Top-left x coordinate in pixels")
    y: int = Field(description="Top-left y coordinate in pixels")
    width: int = Field(description="Width in pixels")
    height: int = Field(description="Height in pixels")

    @property
    def center(self) -> tuple[int, int]:
        return (self.x + self.width // 2, self.y + self.height // 2)


class CandidateElement(BaseModel):
    label: str = Field(description="Descriptive label e.g. 'Sign In button'")
    bbox: BoundingBox
    confidence: float = Field(ge=0, le=1)
    is_chosen: bool = Field(description="True if this is the selected action target")


class DetectedElement(BaseModel):
    label: str
    element_type: str = Field(description="button, input, link, image, text, heading, navigation, modal, form, dropdown")
    bbox: BoundingBox
    confidence: float = Field(ge=0, le=1)
    interactable: bool = True
    accessibility_issues: list[str] = Field(default_factory=list)


class PerceptionResult(BaseModel):
    page_description: str
    page_purpose: str = ""
    has_modal_or_overlay: bool = False
    loading_state: bool = False
    elements: list[DetectedElement] = Field(default_factory=list)


class ActionPlan(BaseModel):
    action_type: Literal[
        "click", "type", "scroll_down", "scroll_up",
        "press_key", "hover", "wait", "go_back", "done"
    ]
    target_element: str | None = None
    coordinates: tuple[int, int] | None = None
    input_text: str | None = None
    key: str | None = None
    seconds: int | None = None
    reasoning: str = Field(description="1-line explanation shown in UI")
    confidence: float = Field(ge=0, le=1)
    candidates: list[CandidateElement] = Field(
        default_factory=list,
        description="Top-3 candidate elements considered"
    )


FRICTION_CATEGORIES = ("navigation", "contrast", "affordance", "copy", "error", "performance", "accessibility")


class CategorizedFriction(BaseModel):
    category: str = Field(description="One of: navigation, contrast, affordance, copy, error, performance, accessibility")
    description: str = Field(description="Specific friction point description")
    severity: Literal["low", "medium", "high", "critical"] = "medium"
    evidence: str = Field(default="", description="Measured data backing this friction point, e.g. '24x18px (min 44px)' or 'term: circle back'")


class EvaluationResult(BaseModel):
    progress_made: bool
    goal_achieved: bool
    page_changed: bool = True
    description: str
    friction_detected: list[str] = Field(default_factory=list)
    friction_items: list[CategorizedFriction] = Field(
        default_factory=list,
        description="Categorized friction points with severity"
    )
    confidence: float = Field(ge=0, le=1)


class ActionResult(BaseModel):
    success: bool
    url_after: str = ""
    error: str | None = None


class NavigationStep(BaseModel):
    step_number: int
    perception: PerceptionResult
    plan: ActionPlan
    action_result: ActionResult
    evaluation: EvaluationResult
    screenshot_before_b64: str = ""
    screenshot_after_b64: str = ""
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    is_recovery: bool = False
    recovery_reason: str | None = None
